training/AUTO_REGISTRATION_OVERVIEW.md

```diff
@@ -0,0 +1,216 @@
+
+Unlike typical Arkouda "message handlers", Chapel procedures annotated with `registerCommand` do not need to adhere to the standard function signature. Instead, they can have formal arguments and return values of any of the following types:
+
+* Chapel scalars supported in Arkouda: `string`, `bool`, `int(?)`, `uint(?)`, `real(?)`, `bigint`
+* Arrays of Chapel scalars
+* Lists of Chapel scalars
+* Homogenous tuples of Chapel scalars
+
+For example, the following command creates a copy of an array and sets the element at the given index to the given value:
+
+```chapel
+@arkouda.registerCommand
+proc foo(a: [?d] ?t, idx: d.rank*int, value: t): [] t {
+    var b = a;
+    b[idx] = value;
+    return b;
+}
+```
+
+Given a pdarray, `arr`, an index tuple that matches `arr`'s rank, and a value that matches `arr`'s dtype, the command is called as follows from the client:
+
+```python
+result = create_pdarray(
+    generic_msg(
+        cmd=f"foo<{arr.dtype},{arr.ndim}>",
+        args={
+            "a": arr,
+            "idx": idx,
+            "value": value
+        },
+    )
+)
+```
+
+Note that the array argument's `dtype` and number of dimensions must be specified in the command. This is because the annotation tells Arkouda to create a separate instantiation of the procedure `foo` for each combination of `dtype` and rank supported by the server — each instantiation is registered as its own command.
+
+<a id="rc-requirements"></a>
+### requirements and notes
+
+* all arguments must have type annotations
+  - generic arguments are allowed as long as their type is specified (i.e., all three arguments in the above example are generic, but its clear that the first is an array with generic domain and element type, the second is a homogenous tuple whose size matches the array's rank, and the third is a scalar whose type matches the array's element type)
+* the return type must be specified
+  - this can be omitted if the procedure doesn't return anything
+* procedures can throw
+  - the server will propagate the error message to the client
+* by default, the prefix in the command name will match the procedure's name
+  - to overwrite this, use `@arkouda.registerCommand("alt-name")` or `@arkouda.registerCommand(name="alt-name")`
+* if more than one array argument is specified, the command will be instantiated for all combinations of both array dtypes and ranks separately (e.g., if the server is configured to support 1D and 2D arrays, and 4 dtypes, then a procedure with two array arguments will be instantiated 64 (2 x 4 x 2 x 4) times).
+  - if the array arguments are intended to always have the same rank, the same dtype, (or both), then `instantiateAndRegister` should be used instead
+  - future work may allow something like the following to avoid instantiating all combinations: `proc foo(a: [?da] ?t, b: [?db] t) where da.rank == db.rank { ... }`
```

Comment:
Just a comment, this would be great to add, as when I tried the two array version of this it crashed b/c of all the combinations.